File: patterns/structural_proxy/structural_proxy_webref.py
class CarAccsess:
    def open(self):
        print('Openning door')

    def close(self):
        print('Close the car door!')


# proxy
class SecuritySystem:
    def __init__(self, door):
        self.door = door

    def open(self, password):
        if self.auntificate(password):
            self.door.open()
        else:
            print("Access denied!")


    def auntificate(self, password):
        return password == 'Ilon'

    def close(self):
        self.door.close() 


if __name__ == "__main__":
    door = SecuritySystem(CarAccsess())

    door.open("Jack")
    door.open("Ilon")
    door.close()

File: patterns/structural_proxy/structural_proxy_webref.js
class CarAccess {
    open() {
        console.log('Opening car door!');
    }

    close() {
        console.log('Closing the car door');
    }
}

// proxy
class SecuritySystem {
    constructor(door) {
        this.door = door;
    }

    open(password) {
        if (this.aunthenticate(password)) {
            this.door.open();
        } else {
            console.log('Access denied!');
        }
    }

    aunthenticate(password) {
        return password === 'Ilon';
    }

    close() {
        this.door.close();
    }
};

// ---
const door = new SecuritySystem(new CarAccess());

door.open('Jack');
door.open('Ilon');
door.close();